### PR TITLE
implementing DB retry logics

### PR DESCRIPTION
**Pull request by kunkun-tang**
when DB runs unstably (or, DB switch), Azkaban today fails to respond generally. Schedule, Execution, Fetch all fail because AZ can not talk to database. In this change, we implement an DB retry logics to enforce retry fetching potential connection, when AZ could not find a stable DB connection.

Still in progress. submit PR for discussion. The main problem is the timeout setting.

Haven't done test yet.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 1]
```diff
@@ -142,6 +146,45 @@ private MySQLBasicDataSource(String host, int port, String dbName,
...
+           * its code hardly.
+           */
+          connection = createDataSource().getConnection();
+          if(connection != null)
+            return connection;
+        } catch (Exception ex) {
+          logger.error("Failed to find DB connection", ex);
+          this.dataSource = null;
+          logger.info("waits 10 seconds and retry. No.Attempt = " + retryAttempt);
+//          sleepMillis(1000L);
```
Comment (ameyamk) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: please remove uncommented code.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 2]
```diff
@@ -142,6 +146,45 @@ private MySQLBasicDataSource(String host, int port, String dbName,
...
+       * we fail the connection immediately.
+       */
+      if (getDataSource() == null) {
+        return createDataSource().getConnection();
+      }
+
+      Connection connection = null;
+      int retryAttempt = 0;
+      this.dataSource = null;
+      while (connection == null && retryAttempt < ServerInternals.MAX_DB_RETRY_COUNT) {
```
Comment (ameyamk) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: This method works all ok - but usually its a good idea to try exponential/ fibonacci backoff in such cases

This is a good read: https://carlosbecker.com/posts/exponential-backoff-java8/
This as well: https://github.com/rholder/guava-retrying

We can keep it simple and not necessarily have to use external dependency - but this is a good example of using exponential backoff

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 3]
```diff
@@ -151,6 +194,21 @@ public boolean allowsOnDuplicateKey() {
     public String getDBType() {
       return "mysql";
     }
+
+    private void sleepMillis(long numMilli) {
+      try {
+        Thread.sleep(numMilli);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
```
Comment (ameyamk) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: why not log error as well?

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 1 above)
Comment (ameyamk) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: can we add unit test to retry logic?

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 4]
```diff
@@ -142,6 +146,45 @@ private MySQLBasicDataSource(String host, int port, String dbName,
...
+      Connection connection = null;
+      int retryAttempt = 0;
+      this.dataSource = null;
+      while (connection == null && retryAttempt < ServerInternals.MAX_DB_RETRY_COUNT) {
+        try {
+          /*
+           * when DB connection could not be fetched here, dbcp library will keep searching until a timeout defined in
+           * its code hardly.
+           */
+          connection = createDataSource().getConnection();
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: How does createDataSource().getConnection() work internaly? Will it establish a new connection or just fetch one from existing connection pool? If establishing a new connection, will it be an issue when db fails, every thread which are querying the DB will try to establish a new connection?

Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: This is a very good question! We indeed have this risk if we keep the today's MysqlDataSource class structure. Let me investigate some simple fix if possible to resolve this issue.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 3 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: Good catch.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 1 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: Will considering how to add test after the code is reviewed to be valid and efficient.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 2 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: I didn't know that. They are good example of good practice!

Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: and we don't need such high retry limit if we use exponential backoff.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 4 above)
Comment (shuzhang1989) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: +1 preventing thundering herd

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 5]
```diff
@@ -142,6 +149,61 @@ private MySQLBasicDataSource(String host, int port, String dbName,
       setTestOnBorrow(true);
     }
 
+    /**
+     * Get a singleton object for MySQL BasicDataSource
+     */
+    public static MySQLBasicDataSource getInstance(String host, int port, String dbName,
+        String user, String password, int numConnections) {
+      if (instance == null) {
+        synchronized (MySQLBasicDataSource.class) {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: This check and set optimization is not safe. See the effective java book.
Please use Guice instead.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 4 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: I am not too concerned as long as there is an upper bound on the threads from AZ hosts that will call this method.

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 6]
```diff
@@ -142,6 +149,61 @@ private MySQLBasicDataSource(String host, int port, String dbName,
...
+          if(connection != null)
+            return connection;
+        } catch (SQLException ex) {
+          this.dataSource = null;
+          logger.error("Failed to find DB connection. waits 1 minutes and retry. No.Attempt = " + retryAttempt, ex);
+        } catch (Exception ex) {
+          logger.error("Exception other than SQLException is thrown", ex);
+          return null;
+        } finally {
+          retryAttempt ++;
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: Will putting this statement inside the catch (SQLException ex) { block have the same effect?

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 7]
```diff
@@ -151,6 +213,13 @@ public boolean allowsOnDuplicateKey() {
     public String getDBType() {
       return "mysql";
     }
+
+    /*
+     * get the parent class's dataSource
+     */
+    private DataSource getDataSource() {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: What's the benefit of having this wrapper?

**Comment by kunkun-tang** [comment]
#1021 reimplements this feature.

**Pull request closed without merging**